exchanges/gateio/ws_ob_update_manager_test.go

```diff
@@ -102,49 +113,121 @@ func TestSyncOrderbook(t *testing.T) {
+	require.NoError(t, err)
+	cache.updates = []pendingUpdate{{update: &orderbook.Update{Pair: pair, Asset: asset.USDTMarginedFutures, UpdateID: math.MaxInt64}}}
+	err = cache.SyncOrderbook(t.Context(), e, pair, asset.USDTMarginedFutures, 0, time.Second)
+	require.ErrorIs(t, err, orderbook.ErrOrderbookInvalid)
+}
 
-		expectedLimit := 20
-		if a == asset.Spot {
-			expectedLimit = 100
-		}
+func TestExtractOrderbookLimit(t *testing.T) {
+	t.Parallel()
+	e := new(Exchange)
+	require.NoError(t, testexch.Setup(e), "Setup must not error")
+	cache := &updateCache{}
+
+	_, err := cache.extractOrderbookLimit(e, 1337)
+	require.ErrorIs(t, err, asset.ErrNotSupported)
+
+	_, err = cache.extractOrderbookLimit(e, asset.Spot)
+	require.ErrorIs(t, err, subscription.ErrNotFound)
+
+	err = e.Websocket.AddSubscriptions(nil, &subscription.Subscription{Channel: subscription.OrderbookChannel, Interval: kline.Interval(time.Millisecond * 420)})
+	require.NoError(t, err)
 
-		b, err := e.Websocket.Orderbook.GetOrderbook(pair, a)
+	_, err = cache.extractOrderbookLimit(e, asset.Spot)
+	require.ErrorIs(t, err, errInvalidOrderbookUpdateInterval)
+
+	err = e.Websocket.RemoveSubscriptions(nil, &subscription.Subscription{Channel: subscription.OrderbookChannel, Interval: kline.Interval(time.Millisecond * 420)})
+	require.NoError(t, err)
+
+	// Add dummy subscription so that it can be matched and a limit/level can be extracted for initial orderbook sync spot.
+	err = e.Websocket.AddSubscriptions(nil, &subscription.Subscription{Channel: subscription.OrderbookChannel, Interval: kline.HundredMilliseconds})
+	require.NoError(t, err)
+
+	for _, tc := range []struct {
+		asset asset.Item
+		exp   uint64
+	}{
+		{asset: asset.Spot, exp: 100},
+		{asset: asset.USDTMarginedFutures, exp: futuresOrderbookUpdateLimit},
+		{asset: asset.CoinMarginedFutures, exp: futuresOrderbookUpdateLimit},
+		{asset: asset.DeliveryFutures, exp: deliveryFuturesUpdateLimit},
+		{asset: asset.Options, exp: optionOrderbookUpdateLimit},
+	} {
+		limit, err := cache.extractOrderbookLimit(e, tc.asset)
 		require.NoError(t, err)
-		require.Len(t, b.Bids, expectedLimit)
-		require.Len(t, b.Asks, expectedLimit)
+		require.Equal(t, tc.exp, limit)
 	}
 }
 
+func TestWaitForUpdate(t *testing.T) {
+	t.Parallel()
+
+	cache := &updateCache{
+		updates: []pendingUpdate{
+			{update: &orderbook.Update{Pair: currency.NewBTCUSD(), Asset: asset.Spot, UpdateID: 1337, AllowEmpty: true, UpdateTime: time.Now()}},
+		},
+	}
+
+	err := cache.waitForUpdate(t.Context(), 1337)
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithDeadline(t.Context(), time.Now())
+	defer cancel()
+	err = cache.waitForUpdate(ctx, 1338)
+	require.ErrorIs(t, err, context.DeadlineExceeded)
+
+	cache.ch = make(chan int64, 1) // Reset channel to avoid deadlock
+	var wg sync.WaitGroup
+	wg.Go(func() {
+		err = cache.waitForUpdate(t.Context(), 1338)
+	})
+	cache.ch <- 1338
+	wg.Wait()
+	require.NoError(t, err)
```

Comment:
For sake of consistency, we should stop using `require`  at the end of the unit test.